identity/identity.go

```diff
@@ -103,66 +64,115 @@ func GetIdentityHeader(ctx context.Context) string {
 	return ""
 }
 
-func checkHeader(id *XRHID, w http.ResponseWriter) error {
+var (
+	ErrMissingIdentity      = errors.New("missing x-rh-identity header")
+	ErrDecodeIdentity       = errors.New("unable to b64 decode x-rh-identity header")
+	ErrInvalidOrgIdIdentity = errors.New("x-rh-identity header has an invalid or missing org_id")
+	ErrMissingIdentityType  = errors.New("x-rh-identity header is missing type")
+)
 
+// CheckBasePolicy performs semantic identity check and returns nil for valid values or
+// an error instead.
+func CheckBasePolicy(id *XRHID) error {
 	if (id.Identity.Type == "Associate" || id.Identity.Type == "X509") && id.Identity.AccountNumber == "" {
 		return nil
 	}
 
 	if id.Identity.OrgID == "" && id.Identity.Internal.OrgID == "" {
-		return doError(w, 400, "x-rh-identity header has an invalid or missing org_id")
+		return ErrInvalidOrgIdIdentity
 	}
 
 	if id.Identity.Type == "" {
-		return doError(w, 400, "x-rh-identity header is missing type")
+		return ErrMissingIdentityType
 	}
 
 	return nil
 }
 
-// EnforceIdentity extracts the X-Rh-Identity header and places the contents into the
-// request context.  If the Identity is invalid, the request will be aborted.
-func EnforceIdentity(next http.Handler) http.Handler {
-	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
-		rawHeaders := r.Header["X-Rh-Identity"]
+// If org_id is not defined at the top level, use the internal one. See:
+// https://issues.redhat.com/browse/RHCLOUD-17717
+func topLevelOrgIDFallback(identity *XRHID) {
+	if identity.Identity.OrgID == "" && identity.Identity.Internal.OrgID != "" {
+		identity.Identity.OrgID = identity.Identity.Internal.OrgID
+	}
+}
 
-		// must have an x-rh-id header
-		if len(rawHeaders) != 1 {
-			doError(w, 400, "missing x-rh-identity header")
-			return
-		}
+// DecodeIdentity returns identity value decoded from a base64 JSON encoded
+// string.
+//
+// To put identity into a context, use WithIdentity or DecodeIdentityCtx functions.
+func DecodeIdentity(header string) (XRHID, error) {
+	if header == "" {
+		return XRHID{}, ErrMissingIdentity
+	}
 
-		// must be able to base64 decode header
-		idRaw, err := base64.StdEncoding.DecodeString(rawHeaders[0])
-		if err != nil {
-			doError(w, 400, "unable to b64 decode x-rh-identity header")
-			return
-		}
+	idRaw, err := base64.StdEncoding.DecodeString(header)
+	if err != nil {
+		return XRHID{}, ErrDecodeIdentity
+	}
 
-		var jsonData XRHID
-		err = json.Unmarshal(idRaw, &jsonData)
-		if err != nil {
-			log.Printf("unable to unmarshal x-rh-identity header: %s", err)
-			doError(w, 400, "x-rh-identity header does not contain valid JSON")
-			return
-		}
+	var id XRHID
+	err = json.Unmarshal(idRaw, &id)
+	if err != nil {
+		return XRHID{}, fmt.Errorf("x-rh-identity header does not contain valid JSON: %w", err)
```

Comment:
Could the errors be organized into a type like `IdentityError` and this become a
`IdentityUnmarshalError` error?

Reply:
Wrapping multiple errors is only supported from Go 1.22, this library is 1.16+ and I do not want to change that in this PR. I am not a huge fan of go.uber.org/multierr. I think it is a good idea anyway, so let me do it in a way that this will be ready for multi-error in Go 1.22:

```
	// TODO: use multiple error wrapping once Go 1.22 compatibility is reached
	return XRHID{}, fmt.Errorf("%w: %s", ErrUnmarshalIdentity, err.Error())
```

For introducing a new type, I do not see what it brings to the table. I am keeping using the standard library `error` type.

Reply:
_I edited the above comment, I was wrong about the Go version which introduced multi-error wrapping. Time flies! :-)_

Reply:
Alternatively, I can bump to 1.20 and take advantage of multi-error wrapping in this case. Since 1.20 is already in ubi8/ubi9 and because this is a v2 version bump, I think we could afford such bump.

The only concern about that is identity is not the only library part of this repository, there is also logging and request_id. We don’t use these in our app, both only export a dozen of lines. I can imagine squeezing some incompatible v2 changes into these packages as well:

* unexport https://github.com/RedHatInsights/platform-go-middlewares/blob/master/request_id/request_id.go#L21
* upgrade to aws sdk v2 https://github.com/RedHatInsights/platform-go-middlewares/blob/master/logging/cloudwatch/hook.go

Reply:
This seems reasonable to me. This package is intended to be used in server applications. Those applications can readily increment the version of Go they're using, and if 1.20 is available in ubi8 and ubi9, it's a reasonable baseline.